client/matching/loadbalancer.go

```diff
@@ -135,6 +140,144 @@ func (lb *defaultLoadBalancer) pickPartition(
 		return taskQueue.GetName()
 	}
 
-	n := max(1, nPartitions(nsName.String(), tqName.BaseNameString(), taskQueueType))
+	n := max(1, lb.nWritePartitions(nsName.String(), tqName.BaseNameString(), taskQueueType))
 	return tqName.WithPartition(rand.Intn(n)).FullName()
 }
+
+// PickReadPartition picks a partition for poller to poll task from, and keeps load balanced between partitions.
+// Caller is responsible to call pollToken.Release() after complete the poll.
+func (lb *defaultLoadBalancer) PickReadPartition(
+	namespaceID namespace.ID,
+	taskQueue taskqueuepb.TaskQueue,
+	taskQueueType enumspb.TaskQueueType,
+	forwardedFrom string,
+) *pollToken {
+	if forwardedFrom != "" || taskQueue.Kind == enumspb.TASK_QUEUE_KIND_STICKY {
+		// no partition for sticky task queue and forwarded request
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	parsedName, err := tqname.Parse(taskQueue.GetName())
+	if err != nil || err == nil && !parsedName.IsRoot() {
+		// parse error or partition already picked, use as-is
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	tqlb := lb.getTaskQueueLoadBalancer(namespaceID, parsedName, taskQueueType)
+
+	// For read path it's safer to return global default partition count instead of root partition, when we fail to
+	// map namespace ID to name.
+	var partitionCount = dynamicconfig.GlobalDefaultNumTaskQueuePartitions
+
+	namespaceName, err := lb.namespaceIDToName(namespaceID)
+	if err == nil {
+		partitionCount = lb.nReadPartitions(string(namespaceName), parsedName.BaseNameString(), taskQueueType)
+	}
+
+	return tqlb.pickReadPartition(partitionCount, lb.forceReadPartition())
+}
+
+func (lb *defaultLoadBalancer) getTaskQueueLoadBalancer(
+	namespaceID namespace.ID, parsedName tqname.Name, tqType enumspb.TaskQueueType,
+) *tqLoadBalancer {
+	key := taskQueueKey{NamespaceID: namespaceID, Name: parsedName, Type: tqType}
+
+	lb.lock.RLock()
+	tqlb, ok := lb.taskQueueLBs[key]
+	lb.lock.RUnlock()
+	if ok {
+		return tqlb
+	}
+
+	lb.lock.Lock()
+	tqlb, ok = lb.taskQueueLBs[key]
+	if !ok {
+		tqlb = newTaskQueueLoadBalancer(key)
+		lb.taskQueueLBs[key] = tqlb
+	}
+	lb.lock.Unlock()
+	return tqlb
+}
+
+func newTaskQueueLoadBalancer(key taskQueueKey) *tqLoadBalancer {
+	return &tqLoadBalancer{
+		taskQueue: key,
+		lock:      sync.Mutex{},
```

Comment:
don't need this line, the zero value is fine

Reply:
addressed in #4981